### PR TITLE
CD resources and datasources

### PR DESCRIPTION
**Pull request by gmarjoram**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: ibm/provider/provider.go [hunk 1]
```diff
@@ -1006,9 +1039,40 @@ func Provider() *schema.Provider {
 			"ibm_en_subscription_ios":     eventnotification.ResourceIBMEnFCMSubscription(),
 			"ibm_en_subscription_chrome":  eventnotification.ResourceIBMEnFCMSubscription(),
 			"ibm_en_subscription_firefox": eventnotification.ResourceIBMEnFCMSubscription(),
-			"ibm_en_subscription_slack":   eventnotification.ResourceIBMEnSlackSubscription(),
```
Comment (hkantare) [inline ibm/provider/provider.go:1009]: Don't remove the other serivce resources/datasources

File: website/docs/d/cd_toolchain.html.markdown [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+```
+
+## Argument Reference
+
+Review the argument reference that you can specify for your data source.
+
+* `toolchain_id` - (Required, Forces new resource, String) ID of the toolchain.
+  * Constraints: The maximum length is `36` characters. The minimum length is `36` characters. The value must match regular expression `/^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}$/`.
+
+## Attribute Reference
```
Comment (hkantare) [inline website/docs/d/cd_toolchain.html.markdown:28]: Attributes are computed variables we can't have Required/Optional for those. same for all docs
Eg * `account_id` - (String) Account ID where toolchain can be found.

File: website/docs/r/tekton_pipeline.html.markdown [hunk 3]
```diff
@@ -0,0 +1,226 @@
+---
```
Comment (hkantare) [inline website/docs/r/tekton_pipeline.html.markdown:1]: The names of the files can we have same convention as resource/datasource
Eg rename tekton_pipeline.html,markdown to cd_tekton_pipeline.html,markdown

File: website/docs/r/tekton_pipeline_definition.html.markdown [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+In addition to all argument references listed, you can access the following attribute references after your resource is created.
+
+* `id` - The unique identifier of the tekton_pipeline_definition.
+* `definition_id` - (Required, String) UUID.
+  * Constraints: The maximum length is `36` characters. The minimum length is `36` characters. The value must match regular expression `/^[-0-9a-z]+$/`.
+* `service_instance_id` - (Required, String) UUID.
+  * Constraints: The maximum length is `36` characters. The minimum length is `36` characters. The value must match regular expression `/^[-0-9a-z]+$/`.
+
+## Provider Configuration
```
Comment (hkantare) [inline website/docs/r/tekton_pipeline_definition.html.markdown:54]: This provider configuation may niot be required already we have seperate doc

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,1192 @@
+// Copyright IBM Corp. 2022 All Rights Reserved.
```
Comment (hkantare) [inline ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:1]: file name can we same as resource/datasource naming
->resource_ibm_cd_tekton_pipeline

File: ibm/provider/provider.go (hunk 1 above)
Comment (gmarjoram) [reply ibm/provider/provider.go:1009]: They're not removed. They're further down in the file, I think the git merge put them there. I'll move them to the correct location.

File: ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	return &schema.Resource{
+		ReadContext: DataSourceIBMCdToolchainRead,
+
+		Schema: map[string]*schema.Schema{
+			"toolchain_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "ID of the toolchain.",
+			},
+			"id": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (gmarjoram) [reply ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go:29]: Fix committed

File: ibm/service/cdtoolchain/data_source_ibm_cd_toolchain_tool_bitbucketgit.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+	if err != nil {
+		log.Printf("[DEBUG] GetToolByIDWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("GetToolByIDWithContext failed %s\n%s", err, response))
+	}
+
+	if *getToolByIDResponse.ToolTypeID != "bitbucketgit" {
+		return diag.FromErr(fmt.Errorf("Retrieved tool is not the correct type: %s", *getToolByIDResponse.ToolTypeID))
+	}
+
+	d.SetId(*getToolByIDResponse.ID)
```
[the remark this answers is not in the document]
Comment (gmarjoram) [reply ibm/service/cdtoolchain/data_source_ibm_cd_toolchain_tool_bitbucketgit.go]: Fix committed

File: ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gmarjoram) [reply ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go]: Fix committed

File: ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+	getToolchainByIDResponse, response, err := cdToolchainClient.GetToolchainByIDWithContext(context, getToolchainByIDOptions)
+	if err != nil {
+		log.Printf("[DEBUG] GetToolchainByIDWithContext failed %s\n%s", err, response)
+		return diag.FromErr(fmt.Errorf("GetToolchainByIDWithContext failed %s\n%s", err, response))
+	}
+
+	d.SetId(fmt.Sprintf("%s", *getToolchainByIDOptions.ToolchainID))
+
+	if err = d.Set("id", getToolchainByIDResponse.ID); err != nil {
```
[the remark this answers is not in the document]
Comment (gmarjoram) [reply ibm/service/cdtoolchain/data_source_ibm_cd_toolchain.go]: Fix committed

File: website/docs/d/cd_toolchain.html.markdown (hunk 2 above)
Comment (gmarjoram) [reply website/docs/d/cd_toolchain.html.markdown:28]: We’re using the SDK generator for all our code & doc. Is this :point_up: a bug in the generator or is there something wrong with our swagger.

Do you think we’ll have to manually update each doc ??? (I’m very hesitant to do manual updates)

File: ibm/service/cdtektonpipeline/data_source_ibm_tekton_pipeline.go [hunk 9]
```diff
@@ -0,0 +1,1186 @@
...
+	"github.com/IBM-Cloud/terraform-provider-ibm/ibm/flex"
+	"github.com/IBM/continuous-delivery-go-sdk/cdtektonpipelinev2"
+)
+
+func DataSourceIBMTektonPipeline() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: DataSourceIBMTektonPipelineRead,
+
+		Schema: map[string]*schema.Schema{
+			"id": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/data_source_ibm_tekton_pipeline.go:24]: Just to clarify, is it fine to have "id" as a sub variable? like  here?
```
"toolchain": &schema.Schema{
				Type:        schema.TypeList,
				Computed:    true,
				Description: "Toolchain object.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"id": &schema.Schema{
							Type:        schema.TypeString,
							Computed:    true,
							Description: "UUID.",
						},
						"crn": &schema.Schema{
							Type:        schema.TypeString,
							Computed:    true,
							Description: "The CRN for the toolchain that containing the tekton pipeline.",
						},
					},
				},
			},
```

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go [hunk 10]
```diff
@@ -0,0 +1,1192 @@
...
+					},
+				},
+			},
+			"triggers": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Tekton pipeline triggers list.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:210]: Need clarification please:  Is it fine to have `Computed:  true`, and `optional: true` at the same time? Cause there are cases where a variable might not be exist due to different trigger type.

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go [hunk 11]
```diff
@@ -0,0 +1,1192 @@
...
+	createTektonPipelineOptions := &cdtektonpipelinev2.CreateTektonPipelineOptions{}
+
+	if _, ok := d.GetOk("worker"); ok {
+		workerModel, err := ResourceIBMTektonPipelineMapToWorkerWithID(d.Get("worker.0").(map[string]interface{}))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		createTektonPipelineOptions.SetWorker(workerModel)
+	}
+	if _, ok := d.GetOk("pipeline_id"); ok {
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:479]: got it, will address

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go [hunk 12]
```diff
@@ -0,0 +1,1192 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Standard RFC 3339 Date Time String.",
+			},
+			"created": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Standard RFC 3339 Date Time String.",
+			},
+			"pipeline_definition": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:182]: Need Clarification please: so is it Compute and Optional/Required should never co-exist? But what if a field is computed but not ganranteed to be there?

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go (hunk 5 above)
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:1]: I can manually change them.. I believe

File: website/docs/r/tekton_pipeline.html.markdown (hunk 3 above)
Comment (SidneyJiang) [reply website/docs/r/tekton_pipeline.html.markdown:1]: will address

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline_definition.go [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	replaceTektonPipelineDefinitionOptions.SetPipelineID(parts[0])
+	replaceTektonPipelineDefinitionOptions.SetDefinitionID(parts[1])
+	replaceTektonPipelineDefinitionOptions.SetServiceInstanceID(d.Get("service_instance_id").(string))
+	replaceTektonPipelineDefinitionOptions.SetID(parts[1])
+
+	hasChange := false
+
+	if d.HasChange("pipeline_id") {
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline_definition.go:195]: will remove all if `d.HasChange("pipeline_id") {` blocks

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline_property.go [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	replaceTektonPipelinePropertyOptions.SetPipelineID(parts[0])
+	replaceTektonPipelinePropertyOptions.SetPropertyName(parts[1])
+	replaceTektonPipelinePropertyOptions.SetName(d.Get("name").(string))
+	replaceTektonPipelinePropertyOptions.SetType(d.Get("type").(string))
+
+	hasChange := false
+
+	if d.HasChange("pipeline_id") {
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline_property.go:278]: will address

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go [hunk 15]
```diff
@@ -0,0 +1,1192 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"pipeline_id": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go:43]: will address

File: ibm/service/cdtektonpipeline/data_source_ibm_tekton_pipeline.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (SidneyJiang) [reply ibm/service/cdtektonpipeline/data_source_ibm_tekton_pipeline.go:24]: anyways, at least will rename the root level `id` to `pipeline_id`

File: ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go (hunk 5 above)
Comment (hkantare) [reply ibm/service/cdtektonpipeline/resource_ibm_tekton_pipeline.go]: yes..you need to manually change them

File: website/docs/r/tekton_pipeline_definition.html.markdown (hunk 4 above)
Comment (hkantare) [reply website/docs/r/tekton_pipeline_definition.html.markdown]: May be this we need to discuss with generator team ..We will do and check why they are adding this section

File: website/docs/d/cd_toolchain.html.markdown (hunk 2 above)
Comment (hkantare) [reply website/docs/d/cd_toolchain.html.markdown:30]: yes its a generator bug we will discuss this with the team

**Comment by hkantare** [comment]
fix the conflicts

**Pull request merged**